lib/ex_webrtc/rtp_receiver/report_recorder.ex

```diff
@@ -0,0 +1,201 @@
+  @moduledoc false
+
+  import Bitwise
+
+  alias ExRTCP.Packet.{ReceiverReport, ReceptionReport}
+
+  @max_u32 0xFFFFFFFF
+  @max_u24 0xFFFFFF
+  @max_seq_no 0xFFFF
+  @breakpoint 0x7FFF
+
+  @type t() :: %__MODULE__{
+          sender_ssrc: non_neg_integer(),
+          media_ssrc: non_neg_integer(),
+          clock_rate: non_neg_integer(),
+          lost_packets: MapSet.t(),
+          last_seq_no: {non_neg_integer(), ExRTP.Packet.uint16()},
+          last_report_seq_no: {non_neg_integer(), ExRTP.Packet.uint16()},
+          last_rtp_timestamp: ExRTP.Packet.uint32(),
+          last_timestamp: integer(),
+          last_sr_ntp_timestamp: ExRTP.Packet.uint32(),
+          last_sr_timestamp: integer(),
+          jitter: float(),
+          total_lost: non_neg_integer()
+        }
+
+  @enforce_keys [:sender_ssrc, :media_ssrc, :clock_rate]
+  defstruct [
+              lost_packets: MapSet.new(),
+              last_seq_no: nil,
+              last_report_seq_no: nil,
+              last_rtp_timestamp: nil,
+              last_timestamp: nil,
+              last_sr_ntp_timestamp: 0,
+              last_sr_timestamp: nil,
+              jitter: 0,
+              total_lost: 0
+            ] ++ @enforce_keys
+
+  @doc """
+  Records incoming RTP Packet.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec record_packet(t(), ExRTP.Packet.t(), integer()) :: t()
+  def record_packet(%{last_seq_no: nil} = recorder, packet, time) do
+    # seq_no == {cycle_no, seq_no as in RTP packet}
+    %__MODULE__{
+      recorder
+      | last_seq_no: {0, packet.sequence_number},
+        last_report_seq_no: {0, packet.sequence_number - 1},
+        last_rtp_timestamp: packet.timestamp,
+        last_timestamp: time
+    }
+  end
+
+  def record_packet(recorder, packet, time) do
+    recorder
+    |> record_seq_no(packet.sequence_number)
+    |> record_jitter(packet.timestamp, time)
+  end
+
+  @doc """
+  Records incoming RTCP Sender Report.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec record_report(t(), ExRTCP.Packet.SenderReport.t(), integer()) :: t()
+  def record_report(recorder, sender_report, time) do
+    # we take the middle 32 bits of the NTP timestamp
+    ntp_ts = sender_report.ntp_timestamp >>> 16 &&& @max_u32
+
+    %__MODULE__{recorder | last_sr_ntp_timestamp: ntp_ts, last_sr_timestamp: time}
+  end
+
+  @doc """
+  Creates an RTCP Receiver Report.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec get_report(t(), integer()) :: {ReceiverReport.t(), t()}
+  def get_report(recorder, time) do
+    received =
+      recorder.last_seq_no
+      |> seq_no_diff(recorder.last_report_seq_no)
+      |> min(@max_u24)
+
+    lost =
+      recorder.lost_packets
+      |> MapSet.size()
+      |> min(@max_u24)
+
+    total_lost = min(recorder.total_lost + lost, @max_u24)
```

Comment:
We should count in duplicate packets here? 

```
This number is
      defined to be the number of packets expected less the number of
      packets actually received, where the number of packets received
      includes any which are late or duplicates.  Thus, packets that
      arrive late are not counted as lost, and the loss may be negative
      if there are duplicates.

```

Reply:
Here I was just following what Pion does, if we wanted to include duplicate packets, then the whole `lost_packets` MapSet is unnecessary, we could just count the number of packets received. To be honest, I'm not really sure what is that, I'll just assumed that we shouldn't count duplicate packets, but I'll look into this more.